### DISCUSSION TITLE
Shortcodes?

### DISCUSSION THREAD

**Discussion by dave1quovadimus** [TukuToi/better-wp-plugin-boilerplate#19] [General]
While it appears you have added a great deal of flexibility for shortcodes, the complexity is such that I have no idea where to start to add a simple shortcode. Can you point to any examples as to how to create a shortcode?

**Comment by guyinpv** [comment]
Off the top of my head you could do this.

1) Create a new folder in your plugin called "shortcodes" that would sit alongside the public and admin folders etc.
2) In the `load_dependencies()` function of the main class, require a new file that you placed in the shortcodes folder, such as `shortcodes/class-shortcodes.php`.
3) Setup the new class such as `class pluginprefix_shortcodes { }` or whatever. And then create methods for each shortcode like `shortcode_showname() { }`. Each method would then be responsible for outputting that specific shortcode.
4) In the main plugin class again, in the constructor, it runs `define_admin_hooks()` and then `define_public_hooks()`. Add another one for `define_shortcodes`.
5) Create that `define_shortcodes()` method and then proceed similar to the `define_admin_hooks()` function. You have to instantiate the new  shortcodes class, and then use `$this->loader->add_shortcode(...)` and pass in whatever values it needs. 
6) Each shortcode you create will point to the class method that will receive the arguments and can return the output.

Each time you want a new shortcode, add it to the shortcodes class as a method, then hook it in the `define_shortcodes` method.

Adjust as necessary from there. In your shortcodes folder you can then have other assets, images, extra classes, includes, functions, templating stuff, etc, as needed.

**Comment by smileBeda** [comment]
Hi, also here sorry for the delay..

1. How to add new shortcodes
See https://github.com/TukuToi/better-wp-plugin-boilerplate/blob/e4e2354a8a3063ba19e9db002cd358fcfb78aa27/plugin-name/includes/class-plugin-name.php#L192. That registers a shortcode. You can add new ones or rename that one. Or delete it.
Then see https://github.com/TukuToi/better-wp-plugin-boilerplate/blob/e4e2354a8a3063ba19e9db002cd358fcfb78aa27/plugin-name/public/class-plugin-name-public.php#L103
That defines the callback (that is the actual code) of your ShortCode. Again, if you register new shortcodes then you will add more such callbacks each with some code in it to return whatever it is supposed to return. Or, if you delete the shortcode - you will delete this function. Or if you rename your shortcode, you'll have to adapt the function

2. You can also take an approach as @guyinpv mentioned if you plan to have tons of shortcodes.
I have done something like that here https://github.com/TukuToi/tukutoi-shortcodes/blob/main/public/class-tkt-shortcodes-shortcodes.php
